[ACTIONS] Twilio new actions

js07 opened the pull request:
Adds Twilio actions:

- List Messages
- Get Message
- Delete Message
- List Message Media
- List Calls
- Get Call
- Delete a Call
- Download Recording Media
- List Recording Transcriptions

Resolves #1445

dylburger commented:
- [ ] [Create components to address specific use cases whenever possible](https://pipedream.com/docs/components/guidelines/#component-scope)
- [ ] [Component `key`s should follow the format `app_name_slug-slugified-component-name`](https://pipedream.com/docs/components/guidelines/#component-key-pattern)
- [ ] Components should follow the [standard directory structure](https://pipedream.com/docs/components/guidelines/#folder-structure)
- [ ] [Prefer Node.js client libraries to REST APIs](https://pipedream.com/docs/components/guidelines/#using-apis-vs-client-libraries)
- [ ] When making API requests, [handle pagination](https://pipedream.com/docs/components/guidelines/#pagination) to ensure all data / events are processed
- [ ] [Use `secret` props](https://pipedream.com/docs/components/guidelines/#capturing-sensitive-data) to capture sensitive data
- [ ] Props and methods should be defined in [app files](https://pipedream.com/docs/components/guidelines/#app-files) whenever possible
- [ ] Document methods with [JS Docs](https://pipedream.com/docs/components/guidelines/#methods)
- [ ] [Use `optional` props](https://pipedream.com/docs/components/guidelines/#optional-vs-required-props) whenever possible, and [set a `default` value](https://pipedream.com/docs/components/guidelines/#default-values) where you can
- [ ] Use [async options](https://pipedream.com/docs/components/guidelines/#async-options) to accept user input wherever possible

vellames-turing commented on components/twilio/actions/delete-call/delete-call.js:
```diff
@@ -0,0 +1,21 @@
+const twilio = require("../../twilio.app.js");
```
Could you please convert all the files related to Twilio from Common `Modules` to Ecma Script? It would be nice if you can convert the sources as well.
To do this you should change the file extension from `js` to `mjs`, change the `require` to `import` and change the `module.exports` to `export default`
e.g:

```suggestion
import twilio from "../../twilio.app.mjs";

export default {
  key: "twilio-delete-call",
  name: "Delete Call",
  description: "Remove a call record from your account",
  version: "0.0.1",
  type: "action",
  props: {
    twilio,
    sid: {
      propDefinition: [
        twilio,
        "sid",
      ],
    },
  },
  async run() {
    return await this.twilio.deleteCall(this.sid);
  },
};
```

vellames-turing commented on components/twilio/actions/delete-call/delete-call.js:
```diff
@@ -0,0 +1,21 @@
+    },
+  },
+  async run() {
+    return await this.twilio.deleteCall(this.sid);
```
I guess we can remove this `await` since you are just returning the value. Every returning promise inside an async function will be automatically "awaited". So, its redundante
Please check [ESLINT rule ](https://eslint.org/docs/rules/no-return-await). You can check this [StackOverflow](https://stackoverflow.com/questions/44806135/why-no-return-await-vs-const-x-await#:~:text=If%20you%20await%20inside%20of,generally%20won't%20cause%20issues.) as well
e.g.
```suggestion
    return this.twilio.deleteCall(this.sid);
```

vellames-turing commented on components/twilio/actions/get-call/get-call.js:
```diff
@@ -0,0 +1,21 @@
+    },
+  },
+  async run() {
+    return await this.twilio.getCall(this.sid);
```
Could you remove this `await` as well please?
```suggestion
    return this.twilio.getCall(this.sid);
```

vellames-turing commented on components/twilio/actions/download-recording-media/download-recording-media.js:
```diff
@@ -0,0 +1,48 @@
+    // https://www.twilio.com/docs/voice/api/recording#fetch-a-recording-media-file
+    const downloadUrl = uri + this.format;
+    const pipeline = promisify(stream.pipeline);
+    return await pipeline(
```
Maybe we can remove the `await` here as well?

vellames-turing commented on components/twilio/actions/list-calls/list-calls.js:
```diff
@@ -0,0 +1,55 @@
+    },
+  },
+  async run() {
+    return await this.twilio.listCalls({
```
Could you remove this `await` as well, please?

vellames-turing commented on components/twilio/actions/list-message-media/list-message-media.js:
```diff
@@ -0,0 +1,29 @@
+    },
+  },
+  async run() {
+    return await this.twilio.listMessageMedia(this.messageId, {
```
Could you remove this `await` as well, please?

vellames-turing commented on components/twilio/actions/list-recording-transcriptions/list-recording-transcriptions.js:
```diff
@@ -0,0 +1,29 @@
+    },
+  },
+  async run() {
+    return await this.twilio.listRecordingTranscriptions(this.recordingID, {
```
Could you remove this `await` as well, please?

vellames-turing commented on components/twilio/twilio.app.js:
```diff
@@ -82,17 +148,163 @@ module.exports = {
+    //  * the Recording to download
+    //  * @returns
+    //  */
+    // async downloadRecordingMedia(sid) {
```
Could you please remove this commented function?

vellames-turing reviewed (changes requested):
Hey @js07 . I did a first quick review. I have some suggestions but the code looks great. I will do a more detailed review soon

vellames-turing reviewed (changes requested):
I found some minor problems, related to docs.

vellames-turing commented on components/twilio/actions/list-messages/list-messages.js:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    return this.twilio.listMessages({
+      to: to || undefined, // Use `undefined` if `to` is empty because Twilio API doesn't
```
Since you you cant send blank, maybe we can create a helper into `twilio.app.js`to sanitize these values and use here and in all other actions that need to be sanitized?
Something like this:
In `twilio.app.js`

```
...

methods: {
    sanitizeEmptyValues(obj) {
      const data = { ...obj }
      const keys = Object.keys(data);
      for (let i = 0; i < keys.length; i++) {
        if (data[keys] === "") {
          data[keys] = undefined
        }
      }
      return data
    },

...
```

In `list-messages.js` and other actions that need sanitization:
e.g

```
...

return this.twilio.listMessages(this.twilio.sanitizeEmptyValues({
      to,
      from: this.from,
      limit: this.limit,
    }));

...
```
Maybe you can destruct `this.twilio` in order to decrease the line length

js07 replied:
I created a function [`omitEmptyStringValues`](https://github.com/PipedreamHQ/pipedream/blob/e30dea4129bee325e4eb337aea76677222e7b082/components/twilio/utils.mjs#L32) in `utils.mjs` which returns a new object without entries whose value is `""`.

vellames-turing reviewed (changes requested):
@js07 Review finished. Hope this helps :)

js07 commented on components/twilio/actions/list-message-media/list-message-media.js:
```diff
@@ -0,0 +1,29 @@
+module.exports = {
+  key: "twilio-list-message-media",
+  name: "List Message Media",
+  description: "Return a list of media associated with your message",
```
I think the description of **List Message Media** in [the issue](https://github.com/PipedreamHQ/pipedream/issues/1445#issue-951952011) is a typo. The description for the corresponding action on [integromat](https://www.integromat.com/) is "Return a list of media associated with your message."

<img width="300" alt="list-message-media" src="https://user-images.githubusercontent.com/19861096/141600730-0028d963-fb74-4b47-895c-3fa643e49013.png">

js07 commented on components/twilio/actions/list-messages/list-messages.js:
```diff
@@ -0,0 +1,55 @@
+      type: "string",
+      label: "From",
+      description: "Read messages sent from only this phone number or alphanumeric sender ID. Format the phone number in E.164 format with a + and country code (e.g., +16175551212).",
+      optional: true,
```
The `from` parameter can always be a phone number for [Create a Call](https://www.twilio.com/docs/voice/api/call-resource#create-a-call-resource), [Create a Message](https://www.twilio.com/docs/sms/api/message-resource#create-a-message-resource), [Read multiple Calls](https://www.twilio.com/docs/voice/api/call-resource#read-multiple-call-resources), and [Read multiple Messages](https://www.twilio.com/docs/sms/api/message-resource#read-multiple-message-resources) according to the API docs, so I changed the `from` options value to always be `number.phoneNumber`.

js07 commented:
Thanks for the review! 

Changes:
- added link to the docs in the description of each action
- removed redundant `await`s
- added `async options` to `messageId`, `parentCallSid`, `sid`, `recordingId`, and `from` props
- sanitized empty string prop values
- converted Twilio JavaScript files from CommonJS to ESModules
- changed `to`/`from` prop order, removed unused commented code

vellames-turing reviewed (approved):
Looks great to me, thanks for the changes. Approved!

js07 commented:
Tested:
- [x] List Messages | Return a list of messages associated with your account
- [x] Get Message | Return details of a message
- [x] Delete Message | Delete a message record from your account
- [x] List Message Media | Return a list of media associated with your account
- [x] List Calls | Return a list of calls associated with your account
- [x] Get Call | Return call resource of an individual call
- [x] Delete a Call | Remove a call record from your account
- [x] Download Recording Media | Download a recording media file
- [x] List Recording Transcriptions | Return a set of transcriptions available for a recording
